File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.c
#include <string.h>
#include <stdlib.h>

// Simple comparison logic for sorting integers in ascending order
int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int minimumPushes(char* word) {
    int freq[26] = {0};
    
    // Directly scan characters using string pointer offsets
    for (int i = 0; word[i] != '\0'; i++) {
        freq[word[i] - 'a']++;
    }
    
    // Fast sort exactly 26 primitive boundaries
    qsort(freq, 26, sizeof(int), compare);
    
    int totalPushes = 0;
    int keyIndex = 0;
    
    for (int i = 25; i >= 0; i--) {
        if (freq[i] == 0) {
            break;
        }
        
        int pushCost = (keyIndex / 8) + 1;
        totalPushes += freq[i] * pushCost;
        keyIndex++;
    }
    
    return totalPushes;
}

File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.py
class Solution:
    def minimumPushes(self, word: str) -> int:
        freq = [0] * 26
        
        # Linear character scan bypassing intermediate data structures
        for char in word:
            freq[ord(char) - 97] += 1
            
        # In-place primitive list sorting
        freq.sort()
        
        total_pushes = 0
        key_index = 0
        
        # Loop backwards over the 26 entries
        for i in range(25, -1, -1):
            if freq[i] == 0:
                break
                
            push_cost = (key_index // 8) + 1
            total_pushes += freq[i] * push_cost
            key_index += 1
            
        return total_pushes
